src/minimetrics/core.py

```diff
@@ -0,0 +1,424 @@
+        raise NotImplementedError()
+
+    def serialize_value(self) -> Any:
+        raise NotImplementedError()
+
+
+class CounterMetric(Metric[float]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value = 0.0
+
+    def add(self, value: float) -> None:
+        self.value += value
+
+    def serialize_value(self) -> Any:
+        return self.value
+
+
+class GaugeMetric(Metric[float]):
+    __slots__ = ("min", "max", "sum", "count", "last")
+
+    def __init__(self) -> None:
+        self.min = float("inf")
+        self.max = float("-inf")
+        self.sum = 0.0
+        self.count = 0.0
+        self.last = float("nan")
+
+    def add(self, value: float) -> None:
+        self.min = min(self.min, value)
+        self.max = max(self.max, value)
+        self.last = value
+        self.count += 1
+        self.sum += value
+
+    def serialize_value(self) -> Any:
+        # For now, we compress gauges with the last value.
+        return self.last
+
+
+class DistributionMetric(Metric[float]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value: List[float] = []
+
+    @property
+    def current_complexity(self) -> int:
+        return len(self.value)
+
+    def add(self, value: float) -> None:
+        self.value.append(float(value))
+
+    def serialize_value(self) -> Any:
+        return self.value
+
+
+class SetMetric(Metric[Union[str, int]]):
+    __slots__ = ("value",)
+
+    def __init__(self) -> None:
+        self.value: Set[Union[str, int]] = set()
+
+    @property
+    def current_complexity(self) -> int:
+        return len(self.value)
+
+    def add(self, value: Union[str, int]) -> None:
+        self.value.add(value)
+
+    def serialize_value(self) -> Any:
+        def _hash(x: Any) -> int:
+            if isinstance(x, str):
+                return zlib.crc32(x.encode("utf-8")) & 0xFFFFFFFF
+            return int(x)
+
+        return [_hash(x) for x in self.value]
+
+
+METRIC_TYPES: Dict[str, Callable[[], Metric[Any]]] = {
+    "c": CounterMetric,
+    "g": GaugeMetric,
+    "d": DistributionMetric,
+    "s": SetMetric,
+}
+
+
+class Aggregator:
+    ROLLUP_IN_SECONDS = 10.0
+    MAX_COMPLEXITY = 100000
+
+    def __init__(self) -> None:
+        self.buckets: Dict[ComposedKey, Metric[Any]] = {}
+        self._bucket_complexity: int = 0
+        self._lock: Lock = Lock()
+        self._running: bool = True
+        self._flush_event: Event = Event()
+        self._force_flush: bool = False
+        # Thread handling the flushing loop.
+        self._flusher: Optional[Thread] = Thread(target=self._flush_loop)
+        self._flusher.daemon = True
+        self._flusher.start()
+
+    def _flush_loop(self) -> None:
+        # We check without locking these variables, such racy check can lead to problems if we are not careful. The most
+        # important invariant of the system that needs to be maintained is that if running and force_flush are false,
```

Comment:
About this invariant: your implementation of `add()` never checks that `_running` is `True`. It checks for `_flusher`, but there's a race condition between calling `stop()`, the flusher checking the flags, flushing occurring (outside of the lock), and `flusher` being set to `None`. I believe this can be trivially solved by checking for `_running` instead of `_flusher`, but please double-check this.